# Use the dataset named as Cross Sell Insurance.xlsx to build a classification model. This data consists of customers and vehicle data and with the help of input features we need to predict whether the customer will buy the insurance. Response is our target variable (1 is yes and 0 is no).

* 1) What is the accuracy when you use a Random forest?
* 2) What is the accuracy when you use an SVM? How does changing the model affect the accuracy?
* 3) Find the top 4 important factors by using Random Forest. What is your inference?
* 4) Which evaluation metrics is important and why?

In [1]:
import pandas as pd

insurance = pd.read_excel('F:\Skillenable\Data frames\Cross Sell Insurance.xlsx')
insurance.head()

id  Gender  Age  Driving_License  Previously_Insured Vehicle_Age  \
0   1    Male   44                1                   0   > 2 Years   
1   2    Male   76                1                   0    1-2 Year   
2   3    Male   47                1                   0   > 2 Years   
3   4    Male   21                1                   1    < 1 Year   
4   5  Female   29                1                   1    < 1 Year   

  Vehicle_Damage  Annual_Premium  Vintage Score  Response  
0            Yes           40454            217         1  
1             No           33536            183         0  
2            Yes           38294             27         1  
3             No           28619            203         0  
4             No           27496             39         0

In [2]:
insurance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   id                  3999 non-null   int64 
 1   Gender              3999 non-null   object
 2   Age                 3999 non-null   int64 
 3   Driving_License     3999 non-null   int64 
 4   Previously_Insured  3999 non-null   int64 
 5   Vehicle_Age         3999 non-null   object
 6   Vehicle_Damage      3999 non-null   object
 7   Annual_Premium      3999 non-null   int64 
 8   Vintage Score       3999 non-null   int64 
 9   Response            3999 non-null   int64 
dtypes: int64(7), object(3)
memory usage: 312.5+ KB


In [3]:
insurance.isnull().sum()

id                    0
Gender                0
Age                   0
Driving_License       0
Previously_Insured    0
Vehicle_Age           0
Vehicle_Damage        0
Annual_Premium        0
Vintage Score         0
Response              0
dtype: int64

In [4]:
insurance.shape

(3999, 10)

In [5]:
# crete dummy data for catagorical values
cat_col = ['Gender', 'Vehicle_Age', 'Vehicle_Damage']
cat_col

['Gender', 'Vehicle_Age', 'Vehicle_Damage']

In [6]:
insurance_dummy = pd.get_dummies(insurance, columns = cat_col, drop_first= True)
insurance_dummy.head()

id  Age  Driving_License  Previously_Insured  Annual_Premium  \
0   1   44                1                   0           40454   
1   2   76                1                   0           33536   
2   3   47                1                   0           38294   
3   4   21                1                   1           28619   
4   5   29                1                   1           27496   

   Vintage Score  Response  Gender_Male  Vehicle_Age_< 1 Year  \
0            217         1            1                     0   
1            183         0            1                     0   
2             27         1            1                     0   
3            203         0            1                     1   
4             39         0            0                     1   

   Vehicle_Age_> 2 Years  Vehicle_Damage_Yes  
0                      1                   1  
1                      0                   0  
2                      1                   1  
3                      0                   0  
4                      0                   0

In [12]:
# create input(X) and output(Y)
Y = insurance_dummy[['Response']]
X = insurance_dummy.drop(columns=['Response'])

from sklearn.model_selection import train_test_split

X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=47)
len(X_train),len(X_test),len(Y_train),len(Y_test)

(3199, 800, 3199, 800)

## 1) What is the accuracy when you use as Random forest?

In [17]:
# creating random forest model
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=1000,random_state=42)
rf = model.fit(X_train,Y_train)



C:\Users\arun\AppData\Local\Temp/ipykernel_17548/95859436.py:5: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf = model.fit(X_train,Y_train)


In [18]:
# Predicting the model on test data
Y_test

Response  Prediction
1568         1           0
648          0           0
3615         0           0
3734         1           0
2515         0           0
...        ...         ...
1323         0           1
3454         0           0
1057         0           0
2439         0           0
2507         0           0

[800 rows x 2 columns]

In [19]:
Y_test['Prediction'] = model.predict(X_test)
Y_test

Response  Prediction
1568         1           0
648          0           0
3615         0           0
3734         1           0
2515         0           0
...        ...         ...
1323         0           1
3454         0           0
1057         0           0
2439         0           0
2507         0           0

[800 rows x 2 columns]

In [20]:
from sklearn.metrics import confusion_matrix, accuracy_score

print(confusion_matrix(Y_test['Response'],Y_test['Prediction']))

[[685  14]
 [ 91  10]]


In [21]:
print(accuracy_score(Y_test['Response'],Y_test['Prediction']))


0.86875


### accuracy of Random Forest is 86.88%

## 3) Find the top 4 important factors by using Random Forest. What is your inference?

In [22]:
model.feature_importances_

array([0.22057818, 0.19092477, 0.00041744, 0.05003073, 0.19030038,
       0.2171911 , 0.02241607, 0.02078281, 0.01200518, 0.07535333])

In [23]:
    forest_importance = pd.Series(model.feature_importances_,index=X.columns)
forest_importance.sort_values(ascending=False).head(10)

id                       0.220578
Vintage Score            0.217191
Age                      0.190925
Annual_Premium           0.190300
Vehicle_Damage_Yes       0.075353
Previously_Insured       0.050031
Gender_Male              0.022416
Vehicle_Age_< 1 Year     0.020783
Vehicle_Age_> 2 Years    0.012005
Driving_License          0.000417
dtype: float64

## 2) What is the accuracy when you use an SVM? How does changing the model affect the accuracy?


In [24]:
# building SVM 
from sklearn.svm import SVC
model1 = SVC(random_state=47)

svc_model = model1.fit(X_train,Y_train)  # fit train data


C:\Users\arun\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [25]:
# prediction
Y_test['Prediction'] = svc_model.predict(X_test)
Y_test


Response  Prediction
1568         1           0
648          0           0
3615         0           0
3734         1           0
2515         0           0
...        ...         ...
1323         0           0
3454         0           0
1057         0           0
2439         0           0
2507         0           0

[800 rows x 2 columns]

In [26]:
# confusion metrix
from sklearn.metrics import confusion_matrix,accuracy_score
print(confusion_matrix(Y_test['Response'],Y_test['Prediction']))

[[699   0]
 [101   0]]


In [27]:
# accuracy
print(accuracy_score(Y_test['Response'],Y_test['Prediction']))


0.87375


###  Accuracy of SVM is 87.375 %; accuracy of SVM is higher than Random Forest 


In [28]:
# from the result SVM is better than random forest